scipy/spatial/ckdtree/src/build.cxx

```diff
@@ -153,51 +200,120 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
             tmp_mins[d] = fixed_val;
             tmp_maxes[d] = fixed_val;
 
-            return build(self, start_idx, end_idx, tmp_maxes, tmp_mins, _median, _compact);
+            return build(self, start_idx, end_idx, tmp_maxes, tmp_mins, _median, _compact, buflist, workers);
         }
 
+        std::unique_ptr<build_worker> worker;
+
         if (CKDTREE_LIKELY(_compact)) {
-            _less = build(self, start_idx, p, maxes, mins, _median, _compact);
-            _greater = build(self, p, end_idx, maxes, mins, _median, _compact);
+            if(workers > 1) {
+                worker.reset(new build_worker());
```

Comment:
RAII violation

Reply:
Right, I will fix it after the discussions.

The constructor and the run function are still split because I am waiting for some reviews on this:
Should I:
1. Put the `build_worker` instance on the stack, as it was before, (no pointer, need `run()`) or
2. Have a heap allocation to reduce the stack size for the recursion (pointer, c'tor and `run()` are the same)?

I do not see a significant performance difference. In the interest of lower memory overhead of the patch, option 2 would be preferred.

I could also move the `unique_ptr` into `build_worker`, having kind-of a PIMPL. This would make option 2 look like option 1 in `build()`, but come with a nested class. I am still sure which style is preferred. The current code looks too much like C for me to assume that this pattern would be welcomed.

Please choose one option.

Reply:
I added version three, please have a look.

Reply:
If you implement https://github.com/scipy/scipy/pull/14481#discussion_r679274205 then exactly `workers` instances of `build_worker` will be created. At most `ceil(log2(workers))` on a single thread. Therefore creating it on the stack is not an issue.

Reply:
There is another small reason to have it on the heap: The member `buflist_` of  `build_worker` is modified by the child thread, this would lead to false sharing of cache lines. It is actually only touched by the child trice though: when creating the initial buffer and retrieving it and when splicing-in the `buflist_` of the grandchild (the read when iterating before splicing will not cached anyway).

Reply:
Just use `alignas(64)` on the stack variable and it will be on its own cache line.

Reply:
This is not very portable, as it depends on the CPU (e.g. Power9 has 128 byte cache lines), so one would actually need to determine the cache line size of the target CPU, for which I do not know a portable way.

I put it in anyway as it makes the code a bit shorter and the false sharing is not a bog issue due to the infrequent access.